web-api/src/personalization/rerank.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{cmp::Ordering, collections::HashMap, hash::Hash};
+
+use itertools::{izip, Itertools};
+use xayn_ai_coi::{nan_safe_f32_cmp_desc, CoiSystem, UserInterests};
+
+use crate::models::{DocumentId, DocumentTag, PersonalizedDocument};
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub(super) struct Rank(usize);
+
+impl Rank {
+    pub(super) fn from_index(index: usize) -> Rank {
+        Rank(index + 1)
```

Comment:
Informal feedback from pink was that ranks starting with `0` was an misunderstanding.